--- /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_user.c.orig	2020-12-22 23:33:47.318549129 -0600
+++ /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_user.c	2020-12-27 22:19:36.368265099 -0600
@@ -192,6 +192,12 @@
 	inotify_event.mask = inotify_mask_to_arg(fsn_event->mask);
 	inotify_event.wd = event->wd;
 	inotify_event.cookie = event->sync_cookie;
+	/* DinterR extra data */
+	inotify_event.count = fsn_event->extra.count;
+	inotify_event.pos = fsn_event->extra.pos;
+	inotify_event.pid = fsn_event->extra.pid;
+	inotify_event.atime.tv_sec = fsn_event->extra.atime.tv_sec;
+	inotify_event.atime.tv_usec = fsn_event->extra.atime.tv_usec;
 
 	/* send the main event */
 	if (copy_to_user(buf, &inotify_event, event_size))
