--- /home/dinter/csce499/linux-4.19.143/include/uapi/linux/inotify.h.orig	2020-12-19 22:42:28.398784607 -0600
+++ /home/dinter/csce499/linux-4.19.143/include/uapi/linux/inotify.h	2020-12-19 22:58:33.978994859 -0600
@@ -24,6 +24,8 @@
 	__u32		cookie;		/* cookie to synchronize two events */
 	__u32		len;		/* length (including nulls) of name */
 	char		name[0];	/* stub for possible name */
+	size_t		count;
+	loff_t		pos;		
 };
 
 /* the following are legal, implemented events that user-space can watch for */
