--- /home/dinter/csce499/linux-4.19.143/include/linux/fsnotify_backend.h.orig	2020-12-16 11:23:39.044662455 -0500
+++ /home/dinter/csce499/linux-4.19.143/include/linux/fsnotify_backend.h	2020-12-16 15:59:46.955269076 -0500
@@ -125,6 +125,9 @@
 	/* inode may ONLY be dereferenced during handle_event(). */
 	struct inode *inode;	/* either the inode the event happened to or its parent */
 	u32 mask;		/* the type of access, bitwise OR for FS_* event types */
+        /* DinterR extra data */
+        size_t count;
+        loff_t *pos;
 };
 
 /*
