--- /home/dinter/csce499/linux-4.19.143/include/linux/fsnotify.h.orig	2020-09-10 09:41:57.514651845 -0500
+++ /home/dinter/csce499/linux-4.19.143/include/linux/fsnotify.h	2020-12-14 23:37:53.766194851 -0600
@@ -187,6 +187,13 @@
 }
 
 /*
+ * fsnotify_dinterr_access - process file read through dinterr
+ */
+static inline void fsnotify_dinterr_access(struct file *file, size_t count, loff_t *pos)
+{
+}
+
+/*
  * fsnotify_modify - file was modified
  */
 static inline void fsnotify_modify(struct file *file)
