--- /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_user.c.orig	2020-12-19 21:55:08.767131405 -0600
+++ /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_user.c	2020-12-19 23:01:04.026533840 -0600
@@ -192,6 +192,9 @@
 	inotify_event.mask = inotify_mask_to_arg(fsn_event->mask);
 	inotify_event.wd = event->wd;
 	inotify_event.cookie = event->sync_cookie;
+	/* DinterR extra data */
+	inotify_event.count = fsn_event->extra.count;
+	inotify_event.pos = fsn_event->extra.pos;
 
 	/* send the main event */
 	if (copy_to_user(buf, &inotify_event, event_size))
