--- /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_fsnotify.c.orig	2020-12-16 12:23:12.842303961 -0500
+++ /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_fsnotify.c	2020-12-16 17:55:13.018160469 -0500
@@ -64,6 +64,8 @@
 	return event_compare(last_event, event);
 }
 
+/* DinterR Update */
+
 int inotify_handle_event(struct fsnotify_group *group,
 			 struct inode *inode,
 			 u32 mask, const void *data, int data_type,
