--- /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_fsnotify.c.orig	2020-12-22 23:33:47.320549134 -0600
+++ /home/dinter/csce499/linux-4.19.143/fs/notify/inotify/inotify_fsnotify.c	2020-12-23 00:25:12.310156775 -0600
@@ -64,6 +64,7 @@
 	return event_compare(last_event, event);
 }
 
+/* DinterR Update */
 int inotify_handle_event(struct fsnotify_group *group,
 			 struct inode *inode,
 			 u32 mask, const void *data, int data_type,
@@ -74,6 +75,7 @@
 	struct inotify_inode_mark *i_mark;
 	struct inotify_event_info *event;
 	struct fsnotify_event *fsn_event;
+	struct fsnotify_extra_data *fsn_data;
 	int ret;
 	int len = 0;
 	int alloc_len = sizeof(struct inotify_event_info);
@@ -81,13 +83,18 @@
 	if (WARN_ON(fsnotify_iter_vfsmount_mark(iter_info)))
 		return 0;
 
+        fsn_data = NULL;
+	if (data_type == FSNOTIFY_EVENT_INODE_PLUS)
+		fsn_data = (struct fsnotify_extra_data*)data;
+
 	if ((inode_mark->mask & FS_EXCL_UNLINK) &&
-	    (data_type == FSNOTIFY_EVENT_PATH)) {
-		const struct path *path = data;
+            (data_type == FSNOTIFY_EVENT_PATH)) {
+	        const struct path *path = data;
 
-		if (d_unlinked(path->dentry))
-			return 0;
+	        if (d_unlinked(path->dentry))
+		        return 0;
 	}
+
 	if (file_name) {
 		len = strlen(file_name);
 		alloc_len += len + 1;
@@ -118,6 +125,11 @@
 	}
 
 	fsn_event = &event->fse;
+        /* DinterR extra data */
+	if (fsn_data) {
+		fsn_event->extra.count = fsn_data->count;
+		fsn_event->extra.pos = fsn_data->pos;
+	}
 	fsnotify_init_event(fsn_event, inode, mask);
 	event->wd = i_mark->wd;
 	event->sync_cookie = cookie;
