--- /home/dinter/csce499/linux-4.19.143/fs/read_write.c.orig	2020-09-10 14:41:55.247838936 -0500
+++ /home/dinter/csce499/linux-4.19.143/fs/read_write.c	2020-12-12 23:54:49.435650008 -0600
@@ -20,6 +20,7 @@
 #include <linux/compat.h>
 #include <linux/mount.h>
 #include <linux/fs.h>
+#include <linux/module.h>
 #include "internal.h"
 
 #include <linux/uaccess.h>
@@ -451,6 +452,9 @@
 			count =  MAX_RW_COUNT;
 		ret = __vfs_read(file, buf, count, pos);
 		if (ret > 0) {
+                        struct module* dinterr_mod = find_module("DinterR_mod");
+                        if (unlikely(try_module_get(dinterr_mod)))
+                                module_put(dinterr_mod);
 			fsnotify_access(file);
 			add_rchar(current, ret);
 		}
