--- /home/dinter/csce499/linux-4.19.143/fs/nfsd/vfs.c.orig	2020-12-22 23:33:47.315549122 -0600
+++ /home/dinter/csce499/linux-4.19.143/fs/nfsd/vfs.c	2020-12-22 23:33:47.143548716 -0600
@@ -902,7 +902,7 @@
 	if (host_err >= 0) {
 		nfsdstats.io_read += host_err;
 		*count = host_err;
-		fsnotify_access(file);
+		fsnotify_access(file, (size_t)*count, offset);
 		trace_nfsd_read_io_done(rqstp, fhp, offset, *count);
 		return 0;
 	} else {
